File: algorithms/matching.java
public int kmp(String strF, String strC, int pos){
  int fLen=strF.length();
  int cLen=strC.length();
  if(pos<0 || pos>fLen){
    pos=1;
  }
  int i=pos,j=1;

  int[] next=getNext(strC);

  while(i<=fLen&&j<=cLen){
    if(0==j||strF.charAt(i-1)==strC.charAt(j-1)){
    	i++;j++;
    }
    else {
    	j=next[j-1];
    }
  }
  if(j>=cLen) return i-j+1;
  return -1;
}
private int[] getNext(String str){
	  int len=str.length();
	  int[] arr=new int[len];
	  int i=1, j=0;
	  arr[0]=0;
	  while(i<len){
		  if(0==j||str.charAt(i-1)==str.charAt(j-1)){
			  arr[i]=++j;
			  i++;
		  }
		  else{
        //i and j always point to the next ele
			  j=arr[j-1];
		  }
	  }

	  return arr;
	}

File: algorithms/matching.js
// father string & child string, return position, if not found, return -1
function kmp(strF, strC, pos) {
  var lenF=strF.length, lenC=strC.length;
  pos=pos&&pos>0&&pos<lenF?pos:1;
  var i=pos-1, j=0;
  var next=getNext(strC);

  while(i<lenF&&j<lenC) {
    if(-1==j||strF[i]==strC[j]){
      i++;j++;
    } else {
      j=next[j]-1;
    }
  }
  if(j>=lenC) {
    return i-j+1;
  }
  return -1;
}
function getNext(str) {
  var i=1, j=0, arr=[];
  arr[0]=0;
  while(i<str.length) {
    if(0==j||str[i-1]==str[j-1]) {
      arr[i]=++j;
      i++;
    } else {
      j=arr[j]-1;
    }
  }
  return arr;
}
